tests/test_spark.py

```diff
@@ -0,0 +1,123 @@
+    SPARK_AGGREGATED_STAGE_METRICS,
+    SPARK_APPLICATION_DIFF_METRICS,
+    SPARK_APPLICATION_GAUGE_METRICS,
+    SPARK_EXECUTORS_METRICS,
+)
+from granulate_utils.metrics.sampler import BigDataSampler
+from pytest import LogCaptureFixture
+
+from gprofiler.log import get_logger_adapter
+from tests.conftest import build_image
+from tests.utils import wait_container_to_start
+
+# `BigDataSampler` receives a logger as an argument.
+logger = get_logger_adapter("gprofiler_spark_test")
+
+# List that includes all the metrics that are expected to be collected by `BigDataSampler` in SA mode.
+EXPECTED_SA_METRICS_KEYS = [
+    metric
+    for metrics_dict in (
+        SPARK_APPLICATION_GAUGE_METRICS,
+        SPARK_APPLICATION_DIFF_METRICS,
+        SPARK_AGGREGATED_STAGE_METRICS,
+        SPARK_EXECUTORS_METRICS,
+    )
+    for metric in metrics_dict.values()
+]
+
+SPARK_MASTER_HOST = "127.0.0.1"
+
+DISCOVER_INTERVAL_SECS = 5
+DISCOVER_TIMEOUT_SECS = 60
+
+
+@pytest.fixture
+def sparkpi_container(docker_client: DockerClient) -> Container:
+    """
+    This fixture is responsible for running SparkPi application in a container.
+    See `containers/spark/Dockerfile`
+    """
+    spark_image = build_image(docker_client=docker_client, runtime="spark")
+    container = docker_client.containers.run(
+        spark_image,
+        detach=True,
+        mounts=[],
+        network_mode="host",
+        pid_mode="host",
+        environment={"SPARK_MASTER_HOST": SPARK_MASTER_HOST},
+    )
+    wait_container_to_start(container)
+    try:
+        yield container
+    finally:
+        container.stop()
+        container.remove()
+
+
+def _validate_spark_sa_metricssnapshot(snapshot: MetricsSnapshot) -> None:
+    """
+    Validates that the snapshot contains all the expected metrics.
+    """
+    samples = snapshot.samples
+    assert len(samples) != 0, "No samples found in snapshot"
+    metric_keys = [sample.name for sample in samples]
+    for key in EXPECTED_SA_METRICS_KEYS:
+        assert key in metric_keys, f"Metric {key} not found in snapshot"
+
+
+def test_spark_sa_discovered_mode(caplog: LogCaptureFixture, sparkpi_container: Container) -> None:
+    """
+    Validates `BigDataSampler`s' `discover()` and `snapshot()` API's in discover mode.
+    In discover mode we do not know what's the cluster mode and master address.
+    """
+    discover_timeout = DISCOVER_TIMEOUT_SECS
+    caplog.set_level(logging.DEBUG)
+    sampler = BigDataSampler(logger, "", None, None, False)
+    # We want to make the discovery process as close as possible to the real world scenario.
+    while not (discovered := sampler.discover()) and discover_timeout > 0:
+        sleep(DISCOVER_INTERVAL_SECS)
+        discover_timeout -= DISCOVER_INTERVAL_SECS
+    assert discovered, "Failed to discover cluster mode and master address"
+    # Sleeping before calling `snapshot()` to make sure SparkPi application is submitted and recognized by Master.
```

Comment:
This is racy. No way to check for some actual event and then sample? For example, if we can run the app to completion and THEN sample the metrics (and we'd still see them).

Reply:
https://github.com/Granulate/gprofiler/pull/784/files#diff-09cfb6f133292348c331f171e17cced53e5dd31dc6c777a92e2241837b253a25R73-R85

Reply:
querying immediately after we know `SparkPi` is running is also problematic, `Master` might take a moment or two to provision.
exceptions were thrown from time to time (e.g. `_spark_application_metrics()`).
sleeping for 5 seconds after we know `SparkPi` is recognize is the best option IMO